File: src/main/kotlin/com/example/webservice/domains/users/controllers/RoleController.java
package com.example.webservice.domains.users.controllers;


import com.example.webservice.domains.users.models.entities.Role;
import com.example.webservice.domains.users.models.entities.User;
import com.example.webservice.domains.users.models.annotations.CurrentUser;
import com.example.webservice.exceptions.forbidden.ForbiddenException;
import com.example.webservice.exceptions.notfound.UserNotFoundException;
import com.example.webservice.domains.users.services.RoleService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/api/v1/roles")
public class RoleController {
    private final RoleService roleService;

    @Autowired
    public RoleController(RoleService roleService) {
        this.roleService = roleService;
    }

    @GetMapping("")
    private ResponseEntity getMyRoles(@CurrentUser User currentUser) throws ForbiddenException, UserNotFoundException {
        List<Role> roles = this.roleService.findByUser(currentUser.getId());
        return ResponseEntity.ok(roles);
    }

}

File: src/main/kotlin/com/example/webservice/domains/users/controllers/RoleController.kt
package com.example.webservice.domains.users.controllers


import com.example.webservice.domains.users.models.dtos.RoleDto
import com.example.webservice.domains.users.models.mappers.RoleMapper
import com.example.webservice.domains.users.services.RoleService
import com.example.webservice.exceptions.notfound.NotFoundException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/admin/roles")
class RoleController @Autowired constructor(
        val roleService: RoleService,
        val roleMapper: RoleMapper
) {

    @GetMapping("")
    fun getAll(@RequestParam("page", defaultValue = "0") page: Int): ResponseEntity<Any> {
        val roles = this.roleService.findAll(page)
        return ResponseEntity.ok(roles.map { this.roleMapper.map(it) })
    }


    @PostMapping("")
    fun create(@RequestBody dto: RoleDto): ResponseEntity<Any> {
        val role = this.roleService.save(this.roleMapper.map(dto, null))
        return ResponseEntity.ok(this.roleMapper.map(role))
    }

    @PatchMapping("/{id}")
    fun update(@PathVariable id: Long,
               @RequestBody dto: RoleDto): ResponseEntity<Any> {
        val exRole = this.roleService.find(id).orElseThrow { NotFoundException("Could not find role with id: $id") }
        val role = this.roleService.save(this.roleMapper.map(dto, exRole))
        return ResponseEntity.ok(this.roleMapper.map(role))
    }

    @GetMapping("/{id}")
    fun get(@PathVariable id: Long): ResponseEntity<Any> {
        val role = this.roleService.find(id).orElseThrow { NotFoundException("Could not find role with id: $id") }
        return ResponseEntity.ok(this.roleMapper.map(role))
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Long): ResponseEntity<Any> {
        this.roleService.delete(id)
        return ResponseEntity.ok().build();
    }
}
